nixos/modules/hardware/device-tree.nix

```
@@ -4,7 +4,106 @@ with lib;
 
 let
   cfg = config.hardware.deviceTree;
+
+  overlayType = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        description = ''
+          Name of this overlay
+        '';
+      };
+
+      dtsFile = mkOption {
+        type = types.nullOr types.path;
+        description = ''
+          Path to .dts overlay file, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        default = null;
+        example = literalExample "./dts/overlays.dts";
+      };
+
+      dtsText = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = ''
+          Literal DTS contents, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        example = literalExample ''
+          /dts-v1/;
+          /plugin/;
+          / {
+                  compatible = "raspberrypi";
+                  fragment@0 {
+                          target-path = "/soc";
+                          __overlay__ {
+                                  pps {
+                                          compatible = "pps-gpio";
+                                          status = "okay";
+                                  };
+                          };
+                  };
+          };
+        '';
+      };
+
+      dtboFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to .dtbo compiled overlay file.
+        '';
+      };
+    };
+  };
+
+  # this requires kernel package
+  dtbsWithSymbols = pkgs.stdenv.mkDerivation {
+        name = "dtbs-with-symbols";
+        inherit (cfg.kernelPackage) src nativeBuildInputs;
+        patches = map (patch: patch.patch) cfg.kernelPackage.kernelPatches;
+
+        buildPhase = ''
+          substituteInPlace scripts/Makefile.lib \
+            --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
+          make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig}
+          make dtbs
+        '';
+
+        installPhase = ''
+          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs
+        '';
```

For cross-compiling I had to make the following changes.

- The `depsBuildBuild` were necessary or else I got an error stating that `bin/sh: gcc: not found`.
- The `ARCH` defines where necessary or else the `dtbs` target weren't available.

```suggestion
        inherit (cfg.kernelPackage) src nativeBuildInputs depsBuildBuild;
        patches = map (patch: patch.patch) cfg.kernelPackage.kernelPatches;

        buildPhase = ''
          substituteInPlace scripts/Makefile.lib \
            --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
          make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig} ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
          make dtbs  ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
        '';

        installPhase = ''
          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs  ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
        '';
```

---

These changes look sensible to me, they are similar to what is done in the kernel package.

---

Thanks!